src/uvc_ros_driver.cpp

```diff
@@ -40,1372 +40,899 @@
+
+  if (fpga_timestamp == 0) {
+    return false;
+  }
+
+  constexpr uint64_t kMicroSecondsToNanoSeconds = 1e3;
+  const uint64_t fpga_timestamp_64 =
+      static_cast<uint64_t>(fpga_timestamp) * kMicroSecondsToNanoSeconds;
+
+  // only update on image timestamps
+  if (update_translator) {
+    device_time_translator_->update(fpga_timestamp_64, ros::Time::now());
+  }
+
+  if (device_time_translator_->isReadyToTranslate()) {
+    *stamp = device_time_translator_->translate(fpga_timestamp);
+  } else {
+    return false;
+  }
+
+  return true;
+}
+
+CamID uvcROSDriver::extractCamId(uvc_frame_t *frame) {
+  constexpr uint8_t kMaxCams = 10;
+  constexpr size_t kCamIdOffset = 8;
+  constexpr size_t kCamIdLine = 0;
+  constexpr size_t kCamIdShift = 4;
+  constexpr size_t kImagesPerFrame = 2;
+  constexpr size_t kRectOffset = 8;
+
+  const uint8_t raw_id =
+      readUInt8(frame, kCamIdLine, kCamIdOffset) >> kCamIdShift;
+
+  CamID cam_id;
+  if (raw_id < kMaxCams / 2) {
+    cam_id.left_cam_num = kImagesPerFrame * raw_id;
+    cam_id.right_cam_num = cam_id.left_cam_num + 1;
+    cam_id.is_raw_images = true;
+  } else {
+    cam_id.left_cam_num = kImagesPerFrame * (raw_id - kRectOffset);
+    cam_id.right_cam_num = cam_id.left_cam_num;
+    cam_id.is_raw_images = false;
+  }
+
+  return cam_id;
+}
+
+uint8_t uvcROSDriver::extractImuId(uvc_frame_t *frame) {
+  constexpr size_t kImuIdOffset = 8;
+  constexpr size_t kImuIdLine = 0;
+  constexpr uint8_t kImuIdMask = 0x0F;
+
+  return readUInt8(frame, kImuIdLine, kImuIdOffset) & kImuIdMask;
+}
+
+uint8_t uvcROSDriver::extractImuCount(size_t line, uvc_frame_t *frame) {
+  constexpr size_t kImuCountOffset = 9;
+  return readUInt8(frame, line, kImuCountOffset);
+}
+
+bool uvcROSDriver::extractImuData(size_t line, uvc_frame_t *frame,
+                                  sensor_msgs::Imu *msg) {
+  if (!extractAndTranslateTimestamp(line, false, frame, &msg->header.stamp)) {
+    return false;
+  }
+
+  msg->linear_acceleration.x =
+      extractImuElementData(line, ImuElement::AX, frame);
+  msg->linear_acceleration.y =
+      extractImuElementData(line, ImuElement::AY, frame);
+  msg->linear_acceleration.z =
+      extractImuElementData(line, ImuElement::AZ, frame);
+
+  msg->angular_velocity.x = extractImuElementData(line, ImuElement::RX, frame);
+  msg->angular_velocity.y = extractImuElementData(line, ImuElement::RY, frame);
+  msg->angular_velocity.z = extractImuElementData(line, ImuElement::RZ, frame);
+
+  return true;
+}
+
+double uvcROSDriver::extractImuElementData(size_t line, ImuElement element,
+                                           uvc_frame_t *frame) {
+
+  constexpr double kDeg2Rad = 2 * M_PI / 360.0;
+  constexpr double kGravity = 9.807;
+  
+  // Standard IMU
+  //constexpr double kAccScaleFactor = kGravity / 16384.0;
+  //constexpr double kGyrScaleFactor = kDeg2Rad / 131.0;
+  
+  // Scaling factors for ADIS.
+  constexpr double kAccScaleFactor = kGravity / 4000.0;
```

Comment:
I think I once got these, maybe we should double-check. which ADIS are you using btw?